// @fusionauth/cli/dist/utilities/apply/http-client.js
/**
 * HTTP Request Execution Engine for FusionAuth CLI Apply command
 * Handles HTTP communication with FusionAuth API
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/**
 * Default timeout configuration
 */
const DEFAULT_TIMEOUTS = {
    timeoutMs: 30000,
};
/**
 * HTTP Client for executing kickstart requests
 */
export class HTTPClient {
    /**
     * Initialize HTTP client
     * @param baseUrl Base URL of FusionAuth instance (e.g., https://auth.example.com)
     * @param apiKey API key for authorization
     * @param timeoutConfig Optional timeout configuration
     */
    constructor(baseUrl, apiKey, timeoutConfig) {
        this.baseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
        this.apiKey = apiKey;
        this.timeoutConfig = Object.assign(Object.assign({}, DEFAULT_TIMEOUTS), timeoutConfig);
    }
    /**
     * Wait for FusionAuth server to be ready
     * Polls /api/status endpoint until it returns JSON response
     * @param maxAttempts Maximum number of attempts (default: 30)
     * @param delayMs Delay between attempts in milliseconds (default: 4000)
     * @returns true if server is ready, throws error if timeout
     */
    waitForServerReady(maxAttempts = 30, delayMs = 4000) {
        return __awaiter(this, void 0, void 0, function* () {
            for (let attempt = 0; attempt < maxAttempts; attempt++) {
                try {
                    const response = yield this.executeRequest('GET', '/api/status', undefined, undefined, undefined, { timeoutMs: 5000 });
                    // Check if response is JSON (not maintenance mode or proxy error)
                    const contentType = response.contentType || 'application/json';
                    if (response.status === 200 &&
                        contentType.toLowerCase().includes('application/json')) {
                        return true;
                    }
                }
                catch (err) {
                    // Ignore errors, will retry
                }
                // Wait before next attempt (except on last attempt)
                if (attempt < maxAttempts - 1) {
                    yield this.sleep(delayMs);
                }
            }
            throw new Error(`Server failed to become ready after ${maxAttempts} attempts`);
        });
    }
    /**
     * Execute an HTTP request to FusionAuth API
     * @param method HTTP method (POST, PATCH, PUT)
     * @param path API path (e.g., /api/tenant/{id})
     * @param body Request body object
     * @param tenantId Optional tenant ID for X-FusionAuth-TenantId header
     * @param contentType Optional content-type override
     * @param customTimeouts Optional custom timeout settings
     * @returns HTTPResponse with status, headers, and body
     */
    executeRequest(method, path, body, tenantId, contentType, customTimeouts) {
        return __awaiter(this, void 0, void 0, function* () {
            // Ensure path starts with a slash
            const normalizedPath = path.startsWith('/') ? path : `/${path}`;
            const url = `${this.baseUrl}${normalizedPath}`;
            const timeouts = Object.assign(Object.assign({}, this.timeoutConfig), customTimeouts);
            const headers = this.buildHeaders(tenantId, contentType);
            const bodyStr = body ? JSON.stringify(body) : undefined;
            try {
                const controller = new AbortController();
                const timeoutId = setTimeout(() => controller.abort(), timeouts.timeoutMs);
                const response = yield fetch(url, {
                    method,
                    headers,
                    body: bodyStr,
                    signal: controller.signal,
                });
                clearTimeout(timeoutId);
                const responseHeaders = this.parseHeaders(response.headers);
                const responseContentType = response.headers.get('content-type') || 'application/json';
                const responseBody = yield this.parseResponseBody(response);
                return {
                    status: response.status,
                    statusText: response.statusText,
                    headers: responseHeaders,
                    body: responseBody,
                    contentType: responseContentType,
                };
            }
            catch (err) {
                if (err instanceof Error) {
                    if (err.name === 'AbortError') {
                        throw new Error(`Request timeout after ${timeouts.timeoutMs}ms: ${method} ${path}`);
                    }
                    throw new Error(`Request failed: ${err.message}`);
                }
                throw new Error(`Request failed: ${String(err)}`);
            }
        });
    }
    /**
     * Execute a DELETE request
     * @param path API path
     * @param tenantId Optional tenant ID
     * @returns HTTPResponse
     */
    executeDelete(path, tenantId) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.executeRequest('DELETE', path, undefined, tenantId);
        });
    }
    /**
     * Check if a resource exists at the given path
     * @param path API path
     * @param tenantId Optional tenant ID
     * @returns true if resource exists (status 2xx or 3xx), false otherwise
     */
    resourceExists(path, tenantId) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield this.executeRequest('GET', path, undefined, tenantId, undefined, { timeoutMs: 5000 });
                return response.status >= 200 && response.status < 400;
            }
            catch (_a) {
                return false;
            }
        });
    }
    /**
     * Build request headers for API call
     * @param tenantId Optional tenant ID
     * @param contentType Optional content-type override
     * @returns Headers object
     */
    buildHeaders(tenantId, contentType) {
        const headers = {
            'Authorization': this.apiKey,
            'Content-Type': contentType || 'application/json',
            'Accept': 'application/json',
            'User-Agent': 'FusionAuth-CLI-Kickstart/1.0',
        };
        if (tenantId) {
            headers['X-FusionAuth-TenantId'] = tenantId;
        }
        return headers;
    }
    /**
     * Parse response headers into a simple object
     */
    parseHeaders(headers) {
        const result = {};
        headers.forEach((value, key) => {
            result[key.toLowerCase()] = value;
        });
        return result;
    }
    /**
     * Parse response body based on content-type
     */
    parseResponseBody(response) {
        return __awaiter(this, void 0, void 0, function* () {
            const contentType = response.headers.get('content-type') || '';
            if (contentType.includes('application/json')) {
                try {
                    return (yield response.json());
                }
                catch (_a) {
                    return yield response.text();
                }
            }
            return yield response.text();
        });
    }
    /**
     * Sleep for a specified number of milliseconds
     */
    sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
    /**
     * Format a request for logging
     */
    formatRequest(method, path, body, tenantId) {
        const url = `${this.baseUrl}${path}`;
        const tenantInfo = tenantId ? ` [tenant: ${tenantId}]` : '';
        const bodyInfo = body ? ` (${JSON.stringify(body).length} bytes)` : '';
        return `${method} ${url}${tenantInfo}${bodyInfo}`;
    }
    /**
     * Format a response for logging
     */
    formatResponse(response) {
        const statusInfo = `${response.status} ${response.statusText}`;
        const bodySize = typeof response.body === 'string'
            ? response.body.length
            : JSON.stringify(response.body).length;
        return `${statusInfo} (${bodySize} bytes)`;
    }
}
/**
 * Helper class for managing request execution with step tracking
 */
export class StepExecutor {
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * Execute a single step and track metrics
     * @param step The parsed step to execute
     * @returns Execution result with response and timing
     */
    executeStep(step) {
        return __awaiter(this, void 0, void 0, function* () {
            const startTime = Date.now();
            const response = yield this.httpClient.executeRequest(step.substitutedRequest.method, step.substitutedRequest.url, step.substitutedRequest.body, step.substitutedRequest.tenantId, step.substitutedRequest.contentType);
            const durationMs = Date.now() - startTime;
            return { response, durationMs };
        });
    }
    /**
     * Check if response indicates success
     */
    isSuccessResponse(response) {
        return response.status >= 200 && response.status < 300;
    }
    /**
     * Extract error details from response
     */
    extractErrorDetails(response) {
        var _a, _b;
        const statusCode = response.status;
        let category = 'unknown_error';
        let message = `HTTP ${statusCode} ${response.statusText}`;
        switch (statusCode) {
            case 400:
                category = 'invalid_payload';
                break;
            case 401:
            case 403:
                category = 'authentication_failed';
                break;
            case 404:
                category = 'not_found';
                break;
            case 409:
                category = 'resource_conflict';
                break;
            case 500:
            case 502:
            case 503:
                category = 'server_error';
                break;
            default:
                break;
        }
        // Try to extract error message from response body
        if (typeof response.body === 'object' && response.body !== null) {
            const body = response.body;
            if (Array.isArray(body.generalErrors) && body.generalErrors.length > 0) {
                const first = body.generalErrors[0];
                message =
                    typeof first === 'string'
                        ? first
                        : String((_a = first === null || first === void 0 ? void 0 : first.message) !== null && _a !== void 0 ? _a : first);
            }
            else if (body.fieldErrors && typeof body.fieldErrors === 'object' && body.fieldErrors !== null) {
                const fieldErrors = body.fieldErrors;
                const firstField = Object.keys(fieldErrors)[0];
                const errs = firstField ? fieldErrors[firstField] : undefined;
                const firstErr = Array.isArray(errs) ? errs[0] : undefined;
                if (firstField) {
                    message = `${firstField}: ${typeof firstErr === 'string' ? firstErr : String((_b = firstErr === null || firstErr === void 0 ? void 0 : firstErr.message) !== null && _b !== void 0 ? _b : firstErr)}`;
                }
            }
            else if (body.message && typeof body.message === 'string') {
                message = body.message;
            }
        }
        return { category, message };
    }
}
