// @fusionauth/cli/dist/utilities/kickstart/validator.js
/**
 * Validator module for FusionAuth CLI Kickstart command
 * Handles schema validation, structure validation, and variable reference validation
 */
import * as fs from 'node:fs';
import { ErrorCategory, HTTPMethod, } from '../apply/types.js';
import { LineTracker } from '../apply/line-tracker.js';
/**
 * Validates kickstart.json configuration files
 */
export class KickstartValidator {
    /**
     * Validate complete kickstart configuration
     * @param config The kickstart configuration to validate
     * @returns ValidationResult with errors if invalid
     */
    validateConfig(config, lineMap = {}) {
        const errors = [];
        const warnings = [];
        // Type check
        if (!config || typeof config !== 'object') {
            errors.push({
                message: 'Kickstart configuration must be a valid JSON object',
                category: ErrorCategory.SCHEMA_INVALID,
            });
            return { valid: false, errors, warnings };
        }
        const cfg = config;
        // Validate variables (optional)
        if (cfg.variables !== undefined) {
            const variablesError = this.validateVariablesStructure(cfg.variables);
            if (variablesError) {
                errors.push(variablesError);
            }
        }
        // Validate requests (required)
        if (!cfg.requests) {
            errors.push({
                field: 'requests',
                message: 'Kickstart configuration must include a "requests" array',
                category: ErrorCategory.SCHEMA_INVALID,
            });
            return { valid: false, errors, warnings };
        }
        const requestsError = this.validateRequestsStructure(cfg.requests);
        if (requestsError.errors.length > 0) {
            errors.push(...requestsError.errors);
            warnings.push(...requestsError.warnings);
        }
        // If requests are valid, check for variable references
        if (errors.length === 0) {
            const variableRefErrors = this.validateVariableReferences(cfg, lineMap);
            errors.push(...variableRefErrors);
        }
        return {
            valid: errors.length === 0,
            errors,
            warnings,
        };
    }
    /**
     * Validate that the kickstart file exists and is readable
     * @param filePath Path to the kickstart file
     * @returns ValidationResult
     */
    validateFileExists(filePath) {
        const errors = [];
        try {
            if (!fs.existsSync(filePath)) {
                errors.push({
                    message: `Kickstart file not found: ${filePath}`,
                    category: ErrorCategory.FILE_NOT_FOUND,
                });
            }
            else if (!fs.statSync(filePath).isFile()) {
                errors.push({
                    message: `Path is not a file: ${filePath}`,
                    category: ErrorCategory.FILE_NOT_FOUND,
                });
            }
        }
        catch (err) {
            errors.push({
                message: `Cannot read file: ${filePath}`,
                category: ErrorCategory.FILE_NOT_FOUND,
            });
        }
        return {
            valid: errors.length === 0,
            errors,
            warnings: [],
        };
    }
    /**
     * Validate JSON structure of kickstart file
     * @param filePath Path to the kickstart file
     * @returns Parsed config with line numbers if valid, or ValidationResult with errors
     */
    loadAndValidateJSON(filePath) {
        const fileError = this.validateFileExists(filePath);
        if (!fileError.valid) {
            return fileError;
        }
        try {
            const content = fs.readFileSync(filePath, 'utf-8');
            const config = JSON.parse(content);
            const lineNumbers = LineTracker.getArrayLineNumbers(filePath, 'requests');
            return { config, lineNumbers };
        }
        catch (err) {
            const message = err instanceof Error ? err.message : 'Unknown error';
            return {
                valid: false,
                errors: [
                    {
                        message: `Invalid JSON in kickstart file: ${message}`,
                        category: ErrorCategory.SCHEMA_INVALID,
                    },
                ],
                warnings: [],
            };
        }
    }
    /**
     * Validate variables object structure
     */
    validateVariablesStructure(variables) {
        if (typeof variables !== 'object' || variables === null) {
            return {
                field: 'variables',
                message: 'Variables must be an object',
                category: ErrorCategory.SCHEMA_INVALID,
            };
        }
        return null;
    }
    /**
     * Validate requests array structure and individual requests
     */
    validateRequestsStructure(requests) {
        const errors = [];
        const warnings = [];
        if (!Array.isArray(requests)) {
            errors.push({
                field: 'requests',
                message: 'requests must be an array',
                category: ErrorCategory.SCHEMA_INVALID,
            });
            return { errors, warnings };
        }
        if (requests.length === 0) {
            errors.push({
                field: 'requests',
                message: 'requests array cannot be empty',
                category: ErrorCategory.SCHEMA_INVALID,
            });
            return { errors, warnings };
        }
        requests.forEach((request, index) => {
            if (typeof request !== 'object' || request === null) {
                errors.push({
                    field: `requests[${index + 1}]`,
                    message: 'Each request must be an object',
                    category: ErrorCategory.SCHEMA_INVALID,
                });
                return;
            }
            const req = request;
            // Validate method
            if (!req.method || typeof req.method !== 'string') {
                errors.push({
                    field: `requests[${index + 1}].method`,
                    message: 'Each request must have a "method" string property',
                    category: ErrorCategory.SCHEMA_INVALID,
                });
            }
            else if (!Object.values(HTTPMethod).includes(req.method)) {
                errors.push({
                    field: `requests[${index + 1}].method`,
                    message: `Method must be one of: ${Object.values(HTTPMethod).join(', ')}. Got: ${req.method}`,
                    category: ErrorCategory.SCHEMA_INVALID,
                });
            }
            // Validate URL
            if (!req.url || typeof req.url !== 'string') {
                errors.push({
                    field: `requests[${index + 1}].url`,
                    message: 'Each request must have a "url" string property',
                    category: ErrorCategory.SCHEMA_INVALID,
                });
            }
            else if (!req.url.startsWith('/api/')) {
                warnings.push(`Request ${index + 1}: URL should start with "/api/": ${req.url}`);
            }
            // Validate body (optional but should be object if present)
            if (req.body !== undefined && (typeof req.body !== 'object' || Array.isArray(req.body) || req.body === null)) {
                errors.push({
                    field: `requests[${index + 1}].body`,
                    message: 'Body must be an object if provided',
                    category: ErrorCategory.SCHEMA_INVALID,
                });
            }
            // Validate contentType (optional, should be string if present)
            if (req.contentType !== undefined &&
                typeof req.contentType !== 'string') {
                errors.push({
                    field: `requests[${index + 1}].contentType`,
                    message: 'contentType must be a string if provided',
                    category: ErrorCategory.SCHEMA_INVALID,
                });
            }
            // Validate tenantId (optional, should be string if present)
            if (req.tenantId !== undefined && typeof req.tenantId !== 'string') {
                errors.push({
                    field: `requests[${index + 1}].tenantId`,
                    message: 'tenantId must be a string if provided',
                    category: ErrorCategory.SCHEMA_INVALID,
                });
            }
        });
        return { errors, warnings };
    }
    /**
     * Validate that all variable references are defined
     */
    validateVariableReferences(config, lineMap = {}) {
        const errors = [];
        const definedVariables = new Set(Object.keys(config.variables || {}));
        // Add default variables that are always available
        definedVariables.add('FUSIONAUTH_APPLICATION_ID');
        definedVariables.add('FUSIONAUTH_TENANT_ID');
        definedVariables.add('TENANT_MANAGER_ID');
        // Check requests
        config.requests.forEach((request, index) => {
            const variableRefs = this.extractVariableReferences(request);
            variableRefs.forEach((varRef) => {
                var _a;
                // special patterns
                if (varRef === 'UUID()' || varRef === 'DEFAULT_TENANT_ID()' || varRef.startsWith('ENV.')) {
                    return;
                }
                if (!definedVariables.has(varRef)) {
                    errors.push({
                        field: `requests[${index + 1}]`,
                        stepId: `step-${String(index + 1).padStart(5, '0')}`,
                        lineNumber: (_a = lineMap[index]) !== null && _a !== void 0 ? _a : index + 1,
                        message: `Undefined variable: #{${varRef}}`,
                        category: ErrorCategory.VARIABLE_NOT_DEFINED,
                    });
                }
            });
        });
        return errors;
    }
    /**
     * Extract variable references from a request
     * Returns array of variable names (without #{})
     */
    extractVariableReferences(request) {
        const refs = new Set();
        // Check URL
        this.extractVariableReferencesFromString(request.url).forEach((r) => refs.add(r));
        // Check tenantId
        if (request.tenantId) {
            this.extractVariableReferencesFromString(request.tenantId).forEach((r) => refs.add(r));
        }
        // Check body
        if (request.body) {
            this.extractVariableReferencesFromObject(request.body).forEach((r) => refs.add(r));
        }
        return Array.from(refs);
    }
    /**
     * Extract variable references from a string
     * Pattern: #{variableName} or #{UUID()} or #{FUSIONAUTH_*}
     */
    extractVariableReferencesFromString(str) {
        const pattern = /#{([^}]+)}/g;
        const matches = [];
        let match;
        // eslint-disable-next-line no-cond-assign
        while ((match = pattern.exec(str)) !== null) {
            matches.push(match[1]);
        }
        return matches;
    }
    /**
     * Extract variable references from an object (recursively)
     */
    extractVariableReferencesFromObject(obj) {
        const refs = new Set();
        const traverse = (value) => {
            if (typeof value === 'string') {
                this.extractVariableReferencesFromString(value).forEach((r) => refs.add(r));
            }
            else if (typeof value === 'object' && value !== null) {
                if (Array.isArray(value)) {
                    value.forEach((item) => traverse(item));
                }
                else {
                    Object.values(value).forEach((item) => traverse(item));
                }
            }
        };
        traverse(obj);
        return Array.from(refs);
    }
}
